File: python/tools/dashboards/create.py
import json

from api.client import is_error, is_success
from lib.utils.api import get_project_base_url
from schema.tool_inputs import DashboardCreateSchema
from tools.types import Context, Tool, ToolResult


async def create_dashboard_handler(context: Context, params: DashboardCreateSchema) -> ToolResult:
    project_id = await context.get_project_id()
    dashboard_result = await context.api.dashboards(project_id).create(params.data)

    if is_error(dashboard_result):
        raise Exception(f"Failed to create dashboard: {dashboard_result.error}")

    assert is_success(dashboard_result)

    dashboard_data = dashboard_result.data

    dashboard_with_url = {
        **dashboard_data.model_dump(),
        "url": f"{get_project_base_url(project_id, context.api.base_url)}/dashboard/{dashboard_data.id}",
    }

    return ToolResult(content=json.dumps(dashboard_with_url))


def create_dashboard_tool() -> Tool[DashboardCreateSchema]:
    return Tool(
        name="dashboard-create",
        description="""
        - Create a new dashboard in the project.
        - Requires name and optional description, tags, and other properties.
        """,
        schema=DashboardCreateSchema,
        handler=create_dashboard_handler,
    )

File: typescript/src/tools/dashboards/create.ts
import { getProjectBaseUrl } from "@/lib/utils/api";
import { DashboardCreateSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = DashboardCreateSchema;

type Params = z.infer<typeof schema>;

export const createHandler = async (context: Context, params: Params) => {
	const { data } = params;
	const projectId = await context.getProjectId();
	const dashboardResult = await context.api.dashboards({ projectId }).create({ data });

	if (!dashboardResult.success) {
		throw new Error(`Failed to create dashboard: ${dashboardResult.error.message}`);
	}

	const dashboardWithUrl = {
		...dashboardResult.data,
		url: `${getProjectBaseUrl(projectId)}/dashboard/${dashboardResult.data.id}`,
	};

	return { content: [{ type: "text", text: JSON.stringify(dashboardWithUrl) }] };
};

const definition = getToolDefinition("dashboard-create");

const tool = (): Tool<typeof schema> => ({
	name: "dashboard-create",
	description: definition.description,
	schema,
	handler: createHandler,
});

export default tool;
